modules/network/src/main/java/org/apache/ignite/internal/network/serialization/PerSessionSerializationService.java

```diff
@@ -53,15 +54,27 @@
      * Map with merged class descriptors. They are the result of the merging of a local and a remote descriptor.
      * The key in this map is a <b>remote</b> descriptor id.
      */
-    private final ConcurrentMap<Integer, ClassDescriptor> mergedDescriptorMap = new ConcurrentHashMap<>();
+    private final Map<Integer, ClassDescriptor> mergedDescriptorMap = new HashMap<>();
 
     /**
-     * Immutable view over {@link #mergedDescriptorMap}. Used by {@link #serializationService}.
+     * A collection of the descriptors that were sent to the remote node.
      */
-    private final Map<Integer, ClassDescriptor> descriptorMapView = Collections.unmodifiableMap(mergedDescriptorMap);
+    private final Set<Integer> sentDescriptors = new IntOpenHashSet();
```

Comment:
I would suggest to change this type to `IntSet` otherwise you will have a lot of unnecessary boxings